Added some improvements on how multi-aperture daophot ascii databases are read

apodemus opened the pull request:
...e read.

embray commented:
Hi, thanks for the PR.  Is there a specific issue this is referencing?  What is this intended to improve?

apodemus commented:
Hi Erik.  Sorry for the delayed reply.  The code I added is intended to improve the way in which multi-aperture daophot ascii databases are read.  Specifically: I encountered some issues when reading such databases in the case when the list of apertures given in the #K APERTURES keyword may not be complete.  This happens if the string description of the aperture list is longer than the field width of the #K APERTURES field (eg:  The list of apertures "1.0, 1.64, 2.29, 2.93, 3.85" is truncated to "1.0, 1.64, 2.29, 2.93, ").  In this case one has to figure out how many apertures there are based on the file structure if you don't want to explicitly pass the number of apertures to use to the function that does the extraction.  I've implemented an automated way of finding the number of apertures by quickly parsing the first line of data before it's read into the formal data structure.
Hope that's enough clarification :)

hamogu commented:
Thanks for your submission. This looks really useful. I've worked with DAOPHOT a little (the first daophot reader in asciitable was written by @taldcroft and me to read my data) and I have never encountered this problem. Then, I usually did not use multiple apertures.

In astropy, we usually review and discuss the code before we merge it and typically some changes are necessary as a result of that discussion. We also add a test to show that the code works and, more importantly, that adding some piece of code does not break what was there before.
(The last pull request that I did has 35 comments and is now substantially different from what I submitted originally.).

Thus, I would ask you to 
1) add a test, where you read a file that contains a truncated list of apertures and you check that everything was read correctly (look in the astropy/io/ascii/test/ directory for examples of how you can do this) and 
2) make sure that the tests we currently have still work. Your code broke some of those tests. The best way to check this is to run the tests on your machine and push code to github once you know that it works:
See http://astropy.readthedocs.org/en/latest/development/testguide.html

Let us know if you are stuck somewhere and we'll try to help you out, but nobody knows the code you've written better than you, so chances are that you know best how to fix it if it does not work perfectly.

pllim commented on astropy/io/ascii/daophot.py:
```diff
@@ -91,7 +101,24 @@ def get_cols(self, lines):
+        if self.unknown_apertures:
+            #determine the number of apertures in the record from the first data-containing line
+            #in the table given the width specifications in the header.
+            def first_true(iterable, default=False, pred=None):
```
Is a nested function really necessary?

embray replied:
I feel like I'm having deja-vu with this comment right here in this code.  Weird....

pllim replied:
Haha. Was it something I wrote? :scream:

embray replied:
I don't know. I feel like I once made a very similar comment on some similar-looking code that you wrote?  Or it was just deja-vu for no reason.  Glitch in the matrix.

pllim replied:
I never wrote nested function. But I am guilty of occasional `lambda` here and there. Or, yes, the Matrix.

pllim commented:
:+1: on adding a test. Also maybe updated the change log.

indiajoe commented:
Basically all we need to do is to count how many repeated lines are there in the data for finding number of multiple apertures. One can see an example file of multiple apertures in the file astropy/io/ascii/tests/t/daophot3.dat
Would it be neater if we simply count the number of lines  ending with *\  to find the number of multiple apertures?
I feel that way the code will be simpler.

indiajoe commented:
What i have in mind is crudely something like this.
```python
        no_of_extra_rows = 0
        for line in lines:
            if re.match(r'\*\\*$',line):  # Look for lines ending with *\ or simply *
                no_of_extra_rows+=1
                StartedCounting = True
            elif StartedCounting:
                break
```
We can then use this no_of_extra_rows to repeat the last column entry as is done in the present code.

We could then also remove the parts of the existing code which reads the APERTURE keyword to find out the number of apertures.

pllim commented:
:+1: on simpler algorithm. However, priority should be given to whichever that has a faster performance so this can read large files within a reasonable time frame.

indiajoe commented:
@pllim both the methods only scan the first data row to find out the number of apertures. so I don't think they will scale with the size of daophot database.

apodemus commented:
Here's a function that takes the raw lines from file and finds the number of continued ("special") lines.  Just not too sure where to put it in the code.  Logically it should be a method of a `Inputter` class.  However, the inputter knows nothing of the table `meta`, which is where the aperture info needs to go.

    def _first_true_idx(iterable, pred=None, default=None):
        func = lambda x : pred(x[1])        if pred         else func = lambda x : x[1]
        ii = next( filter(func, enumerate(iterable)), default )     #either index-item pair or default
        return ii[0]        if ii   else default

    def search_special_rows(lines, depth=150):
        '''search lines for special continuation character to determine number of continued lines.'''
        
        #find first non-comment line
        ixd1 = _first_true_idx( lines[:depth], pred=re_data_finder.match )
        if ixd1 is None: #no data in lines[:depth]
        return
        
        #find first line ending on special row continuation character '*'
        ixe1 = _first_true_idx( lines[ixd1:depth], pred=re_special_continue.search )
        if ixe1 is None:  #no special lines
        return 
        
        ixe2 = _first_true_idx( lines[ixd1+ixe1:depth], pred=re_last_special.search )
        #if ixe1 is None: #no end of special lines within search depth!  increase search depth
        return ixe2

taldcroft commented:
I should have time Friday to look at this in more detail and make some suggestions.

The first question however is what should the output table look like in the case of multi apertures.  My understanding from quickly looking at daophot3.dat is that the "multi" refers to multiple extraction radius values for each source center.  

Do you want to have:

1. New row for each new aperture radius (so that source-related cols like XCENTER, YCENTER get repeated for each aperture radius).  In this case the table gets longer.
2. Give the per-radius values a new index name in the table, so cols like `RAPERT1 SUM1 AREA1 .. RAPERT2 SUM2 AREA2 ...`

I think that option 1) will be much easier for downstream applications to handle since you don't need to handle dynamic column names.  You can actually deal with these aperture groups quite effectively with Table groups (e.g. `group_by('ID'`).  I think these tables are by nature "small" so I wouldn't worry about repeating data.

indiajoe commented:
I think option 1) will be more natural way to read in the data, 
Obtaining an array of magnitudes for each aperture for doing aperture correction will be easier using Table groups, than looping through column names.

When people generate catlogs from daophot output files, one would prefer to generate tables with only one line per source. But this is a writing issue, so may be a convenience function to collapse grouped data to multiple column might be a better solution to tackle this.

taldcroft commented:
OK, so option 1 will be the plan.

In order to do this you will basically need to write a brand new `Inputter` that can handle the wacky way that DAOphot outputs table lines.  The existing `ContinuationLinesInputter` was hacked enough to handle the basics, but for multi-aperture it just won't work.

The starting point is then:
```
class DaophotInputter(core.BaseInputter):
    def process_lines(self, lines):
        ...
```

Start with the existing code for `process_lines` in `core.ContinuationLinesInputter` for guidance but adapt and hardwire it to work for DAOphot files, including multi-aperture support.  When multi-aperture is detected then make a brand new input line for each aperture, copying the common (aperture-independent) bits at the beginning and the aperture-dependent cols at the end.  In this way the final output will have N_source * N_aperture data lines.

Using this strategy there should not be a need to modify the rest of the DAOphot reader code, because this is doing preprocessing to transform the input into what looks like a single aperture file.

When you think you have something working, you should run the tests locally on Python 2 and Python 3 (and make sure everything passes) prior to pushing to github.  Or if you are just interested in getting comments even though things aren't quite working, then put `[skip ci]` at the end of your last commit to avoid burning Travis and Appveyor time.

apodemus commented:
Hey guys.  It seems the build is failing because of unexpected indentation in the documentation.  Any idea how to debug this?

bsipocz commented on astropy/io/ascii/daophot.py:
```diff
@@ -164,16 +198,107 @@ class DaophotData(core.BaseData):
+    
     Example::
-
       #K MERGERAD   = INDEF                   scaleunit  %-23.7g
```
I think these are the lines with the wrong indent.

bsipocz commented on astropy/io/ascii/daophot.py:
```diff
@@ -189,10 +314,9 @@ class Daophot(core.BaseReader):
     The keywords defined in the #K records are available via the output table
     ``meta`` attribute::
-
       >>> import os
```
These are also incorrectly indented.

bsipocz commented on astropy/io/ascii/daophot.py:
```diff
@@ -204,9 +328,8 @@ class Daophot(core.BaseReader):
+    
     The unit and formats are available in the output table columns::
-
       >>> for colname in data.colnames:
```
and these too.

taldcroft commented:
@apodemus - are you able to build the docs locally?  This should be straightforward with `python setup.py build_sphinx`.  That will make debugging the doc updates faster, letting you make sure docs are building properly before submitting for travis/appveyor build testing.

apodemus commented:
@taldcroft - Thanks for the tip! Managed to track down the offending lines, and everything is building OK locally.  Committing now - hopefully this one pulls through! ;)

apodemus commented:
I'm trying to add a test, but getting stuck.  I've added an example file in astropy/io/ascii/tests/t/daophot4.dat, and added a test function to test_read.py, but it doesn't get propagated when doing the build, so the test fails with 
 FileNotFoundError: [Errno 2] No such file or directory: 't/daophot4.dat'

Is there something else I should be doing for this to work?

hamogu commented:
In the `io.ascii` directory there is a file called `setup_package.py`, which determines with data files (files that do not end in .py) are copied over when you install astropy. You need to add `t/daophot4.dat` to that list.

apodemus commented:
Ok, all this seems to be working now!  Your comments please? :)

bsipocz commented:
I think it needs a rebase first, there are several upstream commits not related no this PR and also repeated commits which makes any review difficult. 

Are you familiar with rebasing? Here is the docs for it http://astropy.readthedocs.org/en/latest/development/workflow/additional_git_topics.html#rebasing-on-trunk
If you need help, I can be around on gitter tomorrow.

apodemus commented:
@bsipocz:  I think I incorporated all the upstream commits before my last commit.  I followed these instructions:  https://help.github.com/articles/syncing-a-fork/.  Do I still need to do a rebase?

bsipocz commented:
Yes, please try to do a rebase. I'll log into gitter later, if you need help. 

The instructions you've linked are for a bit different situation (how to update the master of the fork). In general the workflow suggests that developing a new feature or fixing a bug is best done in a new branch (http://astropy.readthedocs.org/en/latest/development/workflow/development_workflow.html?highlight=rebase#workflow).

apodemus commented:
OK, I'm still pretty new to navigating git, so still learning.  I think I did all my editing in the master branch of my fork.  I did the rebase, and after resolving the conflicts ended up here:
    $ git status
    HEAD detached from refs/heads/master
    nothing to commit, working directory clean
When I do  `git push -f` it tells me that "Everything up-to-date"

bsipocz commented:
Yes, sadly git have a steep learning curve. However it'll become logical with time.

It seems like you might have rebased on your own master. What output do you get for the following?
```
git remote -v
```
it should be something like
```
upstream	git://github.com/astropy/astropy.git (fetch)
upstream	git://github.com/astropy/astropy.git (push)
origin	git@github.com:apodemus/astropy.git (fetch)
origin	git@github.com:apodemus/astropy.git (push)
```

If you have all four, two upstream for the astropy version and two for your fork, then saying the following will synchronize the upstream, too:

```
git fetch --all
```
And now the rebase, doing it interactively will open an editor window. It will recognize the commits of others, and will keep only yours. I would also suggest to remove the duplicates (and maybe squash some of the commits):

```
git rebase -i upstream/master
```
I got the following, working on a copy of your master branch (without the line numbers, I've just put them there to ease this). So I would delete lines 20-38 as they are duplicates, and maybe choose ``fixup`` rather than ``pick`` for lines 6-10, and maybe either ``fixup`` or ``reword`` for the ones with the ``test`` commit message, as it's rather cryptic commit message. If you are finished with it, save the file and exit the text editor. I think this should work smoothly, the rebase didn't raise any conflicts for me.  

```
1 pick c9b9c25 Added some improvements on how multi-aperture daophot ascii databases are read.
2 pick 035c04d Second attempt at better handeling of multi-aperture daophot databases
3 pick 433191a Third attempt at better handeling of multi-aperture daophot databases
4 pick c3d0ab1 Fourth attempt at better handeling of multi-aperture daophot databases
5 pick 0e86561 Minor changes in doc strings and dict comprehension
6 pick fc2de8b Minor changes in doc strings and dict comprehension
7 pick a8af94d Doc string fixes
8 pick a35a78d Doc string fixes
9 pick 7bcbac0 Doc string fixes
10 pick bed7ae7 Docstring fixes
11 pick 582313a Adding a test. + Other cosmetic improvements. [skip ci]
12 pick 55ca3af Adding a test.
13 pick 8ba209d Adding a test.
14 pick 60d45cd test
15 pick e82cf36 test
16 pick 69c46df test
17 pick 37a1c19 test
18 pick dca94e9 test
19 pick fafc1bb test
20 pick 83d9d51 Added some improvements on how multi-aperture daophot ascii databases are read.
21 pick 7536682 Second attempt at better handeling of multi-aperture daophot databases
22 pick f7044f8 Third attempt at better handeling of multi-aperture daophot databases
23 pick 9fae76a Fourth attempt at better handeling of multi-aperture daophot databases
24 pick 683e43b Minor changes in doc strings and dict comprehension
25 pick 7c30ad8 Minor changes in doc strings and dict comprehension
26 pick b4b1641 Doc string fixes
27 pick 9519103 Doc string fixes
28 pick 47887dd Doc string fixes
29 pick 263bf6f Docstring fixes
30 pick 8da26ba Adding a test. + Other cosmetic improvements. [skip ci]
31 pick 764b7b2 Adding a test.
32 pick 68b5260 Adding a test.
33 pick 661bc9c test
34 pick eb20927 test
35 pick 0aec769 test
36 pick 6c0aaea test
37 pick ab54a14 test
38 pick 9ac94dd test
39 pick c849235 test
40 pick d7604d1 test
41 pick 42a170b more test
42 pick 3501374 test bugfixes
43 
44 # Rebase e6ecc9c..3501374 onto e6ecc9c
45 #
46 # Commands:
47 #  p, pick = use commit
48 #  r, reword = use commit, but edit the commit message
49 #  e, edit = use commit, but stop for amending
50 #  s, squash = use commit, but meld into previous commit
51 #  f, fixup = like "squash", but discard this commit's log message
52 #  x, exec = run command (the rest of the line) using shell
53 #
54 # These lines can be re-ordered; they are executed from top to bottom.
55 #
56 # If you remove a line here THAT COMMIT WILL BE LOST.
57 #
58 # However, if you remove everything, the rebase will be aborted.
59 #
60 # Note that empty commits are commented out
```

taldcroft commented:
This is still something a little funky because commits 1-19 appear to be repeated in 20-38.

bsipocz commented:
@taldcroft - yes, but this is the list that comes up with the editor during the rebase. And this is why I suggested to remove lines 20-38. And also squash or fixup those docstring tries and test ones.

taldcroft commented:
@bsipocz - sorry I didn't see your comment and thought that what I was seeing was the result of your suggested rebase.

apodemus commented:
@bsipocz - Thanks for the instructions.  I've done the rebase, and it all seemed to have gone well. However the build is now failing again...
It's not really clear from the error message "The command "$MAIN_CMD $SETUP_CMD" exited with 245." how to go about resolving this.  Any idea what may have caused this?

bsipocz commented:
The travis fail seems to be an unrelated glitch, so I think this is ready for review.

hamogu commented on astropy/io/ascii/daophot.py:
```diff
@@ -1,26 +1,99 @@
+        core.BaseHeader.__init__(self)
+        self.comment = r'\s*#K'
+
+        #self.re_digit_extract = re.compile( '\d+' )
```
Is this comment still needed?

apodemus replied:
No, it seems not.

hamogu commented on astropy/io/ascii/daophot.py:
```diff
@@ -1,26 +1,99 @@
+    """Read the header from a file produced by the IRAF DAOphot routine."""
+    def __init__(self):
+        core.BaseHeader.__init__(self)
+        self.comment = r'\s*#K'
```
There is two ways to set class attributes like `comment`, `re_format`, `re_header_keyword` etc. Your method or setting these in the `__init__` works and that's what we initially did for all classes in io.ascii. However, we found that that makes is more cumbersome to derive something else from this class, so instead we decided that is is preferable to set them (as far as possible) like this:
```
class SomeHeader(core.BaseHeader):
    comment =''
```
See #2812 for more details. In this case it does not matter much because DAOPhot is such a specific format that it's very unlikely that we will derive many other classes from it, but I thought I'll just point that out. Change it or leave it as you wish.

apodemus replied:
I found it useful mostly only when I was debugging the code to check whether the header lines where being read correctly.  I found the information in `meta` attribute hard to read because of the multiply nested dictionaries.  This was a kind of intermediary to display the info.    I think it may be useful to add this to the BaseHeader - that way one could more easily check the content of the header.

hamogu commented on astropy/io/ascii/daophot.py:
```diff
@@ -1,26 +1,99 @@
+
+        self.aperture_values = ()
+
+    def __str__():
```
I'm curious why you found this is needed. I have not seen a `__str__` method in in other readers in `io.ascii` and now I wonder if we should add them in other readers, too. What do you think?

apodemus replied:
I found it useful mostly only when I was debugging the code to check whether the header lines where being read correctly.  The info in the `meta` attribute is hard to read at a glance because of the multiply nested dictionaries.  So this was a kind of intermediary.  I think it may be useful to add this to the BaseHeader - that way one could  more easily check the content of the header.

hamogu commented on astropy/io/ascii/daophot.py:
```diff
@@ -28,126 +101,83 @@ def update_meta(self, lines, meta):
-            List of table lines
-
+        :param lines: list of table lines
+        :returns: list of table Columns
```
In genral in astropy we use numpydoc for docstrings (Like before with a "Parameters"  and a "Returns" block). For historic reasons, some parts of `io.ascii` still use the sphinx standard (like `:param lines: text here`), but in new code the astropy guideline strongly advice to use the numpydoc format.

apodemus replied:
Ok, thanks for the info.  Will revert to the numpydoc format.  It's more readable anyway!

hamogu commented on astropy/io/ascii/daophot.py:
```diff
@@ -165,15 +195,105 @@ class DaophotData(core.BaseData):
 
+    def __init__(self):
+        core.BaseData.__init__(self)
+        self.is_multiline = False
```
If instead `is_multiline = False` is defined just under `start_line = 0` (5 lines above), there there is no need to write an `__init__` at all. Again, 2 lines of code saved ;-)

hamogu commented on astropy/io/ascii/daophot.py:
```diff
@@ -165,15 +195,105 @@ class DaophotData(core.BaseData):
+        #This happens if the string description of the aperture list is longer than the
+        #field width of the #K APERTURES field.
+        #In this case we have to figure out how many apertures there are based on the file
+        #structure.
```
Wow! I knew that DAOPhot files are a pain to parse, but I did not realize that it is this bad! Thanks for stepping through all this!

hamogu commented on astropy/io/ascii/daophot.py:
```diff
@@ -165,15 +195,105 @@ class DaophotData(core.BaseData):
+
+    def search_multiline(self, lines, depth=150):
+        '''search lines for special continuation character to determine number of continued rows in a
+        datablock.  For efficiency, depth gives the upper limit of lines to search.'''
```
Do you know how critical performance is in this case? I rarely read more than a dozen DAOPhot files and I would not care if it takes a few seconds longer per file. On the other hand, my files are not big, with only a few hundred objects or so. On the other hand, I've seen DAOPhot files with close to 100 header lines. How much performance would we loose if we set this to e.g. 1000 or any number so large that we are a factor of a few above the usual number of header lines?

apodemus replied:
Instinctively, I feel like it shouldn't have too great an effect on performance to increase this value of the depth parameter.  I can do a few tests to confirm this.  The databases I dealt with are typically a few hundred thousand lines, which means the majority of the time is spent parsing the main body of the database, not the header!

hamogu replied:
I should say, I don't think that this is a critical point either way. I would not spend the effort to run extra speed tests unless you have a use case where you read thousands of files.

hamogu commented on astropy/io/ascii/misc.py:
```diff
@@ -0,0 +1,124 @@
+misc.py:
+  Collection of useful miscellaneous functions.
+
+:Copyright: None -- Hack away at your leisure!!
```
Would you be willing to put this under the same 3-clause BSD license as the rest of astropy? That's a very permissive license that still allows almost everybody to do everything. It just makes managing copyright issues so much easier if all files in a project use the same license. (And all files in astropy should have a header line to to refer to license.rst in the root directory - unfortunately `io.ascii` still has some legacy code in it, but if you look into any other part of astropy you will see it.

hamogu replied:
@astrofrog Maybe some of these helpers are so useful that they should live in a more general `misc` and not only inside `io.ascii`? What do you think?

apodemus replied:
Sure!

taldcroft replied:
@apodemus - sorry for getting to this very late.  Can you add the line below to the top of `misc.py` and remove the `Copyright: None ...` line.  For the astropy project source code we want to have a uniform license including relevant copyright restrictions.
```
# Licensed under a 3-clause BSD style license - see LICENSE.rst
```

hamogu commented:
@taldcroft I've looked through this. I have to admit that I don't understand what happens in every line, but it passes the tests and looks good to me. There is a lot of really clever iterator magic in there, and you know those things much better than I do, so I leave it to you to look over that.

astrofrog commented:
It looks like this needs rebasing to remove the merge commit - can provide instructions later today.

taldcroft commented:
Closed by #4207.  Thanks @apodemus !
